libslide/src/evaluator_rules/pattern_matching.rs

```diff
@@ -59,47 +64,72 @@ pub fn match_rule(rule: ExprPat, target: Expr) -> Option<Replacements> {
 /// rule applied on a target expression.
 pub struct Replacements {
     map: HashMap<
-        ExprPat, // rule pattern, like #a
-        Expr,    // target expr,  like 10
+        u64,      // pointer to rule pattern, like #a
+        Rc<Expr>, // target expr,             like 10
     >,
 }
 
-impl Transformer<ExprPat, Expr> for Replacements {
+impl Transformer<Rc<ExprPat>, Rc<Expr>> for Replacements {
     /// Transforms a pattern expression into an expression by replacing patterns with target
     /// expressions known by the [`Replacements`].
     ///
     /// This transformation can be used to apply a rule on an expression by transforming the RHS
     /// using patterns matched between the LHS of the rule and the target expression.
     ///
     /// [`Replacements`]: Replacements
-    fn transform(&self, item: ExprPat) -> Expr {
-        match item {
-            ExprPat::VarPat(_) | ExprPat::ConstPat(_) | ExprPat::AnyPat(_) => {
-                match self.map.get(&item) {
-                    // We know about this pattern, replace it with its target expression.
-                    Some(transformed) => transformed.clone(),
-
-                    // A pattern can never be transformed to an expression if it is not replaced!
-                    // TODO: Return an error rather than panicking
-                    None => unreachable!(),
-                }
+    fn transform(&self, item: Rc<ExprPat>) -> Rc<Expr> {
+        fn transform(
+            repls: &Replacements,
+            item: Rc<ExprPat>,
+            cache: &mut HashMap<u64 /* ptr address */, Rc<Expr>>,
+        ) -> Rc<Expr> {
```

Comment:
again this comment is awkward. You already said u64 is the address above in line 67

Reply:
Yeah I'll remove this